src/position.cpp

```diff
@@ -204,89 +215,208 @@ Position& Position::set(const string& fenStr, bool isChess960, StateInfo* si, Th
+  File file = FILE_A;
+  Rank rank = RANK_8;
+
   // 1. Piece placement
-  while ((ss >> token) && !isspace(token))
+  for (;;)
   {
-      if (isdigit(token))
-          sq += (token - '0') * EAST; // Advance the given number of files
+      if (!(ss >> token))
+          return PositionSetError("Invalid FEN. Unexpected end of stream.");
+
+      if (isspace(token))
+          break;
 
+      if (isdigit(token))
+      {
+          const int diff = token - '0';
+          if (diff < 1 || diff > 8)
+              return PositionSetError("Invalid FEN. Invalid number of squares to skip.");
+          file = File(file + diff);
+          if (file > FILE_NB)
+              return PositionSetError("Invalid FEN. Invalid file reached.");
+      }
       else if (token == '/')
-          sq += 2 * SOUTH;
+      {
+          if (file != FILE_NB)
+              return PositionSetError("Invalid FEN. Trying to end rank when not at the end of it.");
+
+          --rank;
+          file = FILE_A;
 
-      else if ((idx = PieceToChar.find(token)) != string::npos) {
+          if (rank < RANK_1)
+              return PositionSetError("Invalid FEN. Invalid rank reached.");
+      }
+      else
+      {
+          const size_t idx = PieceToChar.find(token);
+          if (idx == string::npos)
+              return PositionSetError(std::string("Invalid FEN. Invalid piece: ") + std::string(1, token));
+          if (++numPieces > 32)
+              return PositionSetError("Invalid FEN. More than 32 pieces on the board.");
+
+          const Square sq = make_square(file, rank);
           put_piece(Piece(idx), sq);
-          ++sq;
+
+          ++file;
+          if (file > FILE_NB)
+              return PositionSetError("Invalid FEN. Invalid file reached.");
       }
   }
 
+  if (rank != RANK_1 || file != FILE_NB)
+      return PositionSetError("Invalid FEN. Board state encoding ended but cursor not at end.");
+
+  {
+      const int wPawns = count<PAWN>(WHITE);
+      const int bPawns = count<PAWN>(BLACK);
+      if (wPawns > 8)
+          return PositionSetError("Invalid FEN. WHITE has more than 8 pawns.");
+      if (bPawns > 8)
+          return PositionSetError("Invalid FEN. BLACK has more than 8 pawns.");
+
+      const int wAdditionalKnights = std::max((int)count<KNIGHT>(WHITE) - 2, 0);
+      const int bAdditionalKnights = std::max((int)count<KNIGHT>(BLACK) - 2, 0);
+      const int wAdditionalBishops = std::max((int)count<BISHOP>(WHITE) - 2, 0);
+      const int bAdditionalBishops = std::max((int)count<BISHOP>(BLACK) - 2, 0);
+      const int wAdditionalRooks = std::max((int)count<ROOK>(WHITE) - 2, 0);
+      const int bAdditionalRooks = std::max((int)count<ROOK>(BLACK) - 2, 0);
+      const int wAdditionalQueens = std::max((int)count<QUEEN>(WHITE) - 1, 0);
+      const int bAdditionalQueens = std::max((int)count<QUEEN>(BLACK) - 1, 0);
+      if (wAdditionalKnights + wAdditionalBishops + wAdditionalRooks + wAdditionalQueens > 8 - wPawns)
```

Comment:
If one defines `8 - wPawns` as `maxPromotions`, the latter could be set to zero if all opposing pawns are still on their original rank. (plus some other cases, where they form an impenetrable wall, that could never have been overcome by an opposing pawn). But these kind of complicated checks may already go beyond the original scope of this PR.